Adding a *.domain.tld possibilty

pyreweb2 opened the pull request:
I read quickly RFC's and CORS website, it doesn't allow nor disallow this; I needed it, it may be useful to somebody else.

Changes proposed in this pull request:

 * Allow to have a `'allow_origins' => ['https://*.example.com']` to match all subdomains from example.com

pyreweb2 committed:
Adding a *.domain.tld possibilty

Allow to have a `'allow_origins' => ['https://*.example.com']` to match all subdomains

Jalle19 reviewed (changes requested):
I think you added this snippet to the wrong method? It should probably be in `isOriginAllowed()`.

Jalle19 commented on src/CorsService.php:
```diff
@@ -379,6 +379,42 @@ protected function isMethodAllowed($method)
             return true;
         }
+        
+        if (preg_match('#\*\.#', implode('|', $this->allowOrigins)) !== false) {
```
`allow_origins` can be an array so there's no point in splitting strings

Jalle19 reviewed (changes requested):
Can you use a library like `league/uri` to break down the URL instead of this manual stuff?

Pyreweb commented:
> I think you added this snippet to the wrong method? It should probably be in isOriginAllowed().

Yup, sorry, I copy-pasted too quickly.

> Can you use a library like league/uri to break down the URL instead of this manual stuff?

Oh, good idea ! I'll look after it.

> Would it perhaps make more sense to add regex support like here: nelmio/NelmioCorsBundle#65 ?

Well, if you prefer, but I liked the fact that it still looked like normal URIs in the config file.

pyreweb2 committed:
Adding a *.domain.tld possibilty (pt2)

Moved the added part to the correct function

Jalle19 commented:
I think it would be easier to implement regular expressions than parsing the URL for wildcards etc., plus it can be more flexible, e.g. you may only want to allow `app1.example.com, `app2.example.com` up to `app99.example.com`

pyreweb2 committed:
Adding league/uri

pyreweb2 committed:
Using league/uri to parse URI

Pyreweb commented:
I added and used league/uri. I must admit, it's cleaner 😄 
It seems to require and use PHP 7+, and this require 5.6+. You may want to upgrade the PHP version requirement, as the last Lumen use 7+ ?


> I think it would be easier to implement regular expressions than parsing the URL

I'm not a big fan of RegExp in fact, but if you think it's for the better...
Do you sees it as a `foreach (domain) if (preg_match(domain, origin)) return true` ?
If so, I'll close this and make a new PR.

Jalle19 commented:
It should not require a version that is incompatible with PHP 5.6 if you run composer require league/uri

Pyreweb commented:
The 4.2.3 is the last compatible with PHP <7, but it's not supported anymore by PHPLeague since 06/08/2017. We should not use that.

Jalle19 commented:
I don't think it's a dealbreaker if it's supported or not as long as it works for this particular use case.

pyreweb2 committed:
Lowering league/uri version for PHP 5.6

pyreweb2 committed:
Downgrading league/uri

It's not the same syntax.

Aaaand, I corrected my use of preg_match, which return 0 instead of false.

Pyreweb commented:
So be it !
I've made the edits.

cniska commented on composer.json:
```diff
@@ -10,7 +10,8 @@
     "require": {
-        "php": ">=5.6"
+        "php": ">=5.6",
+		"league/uri": "4.2.3"
```
Please fix the indentation.

cniska commented on src/CorsService.php:
```diff
@@ -356,6 +357,43 @@ protected function isOriginAllowed($origin)
+					}
+				}
+			}
+		}
```
Fix the indentation here as well. 

In addition to that this method looks a bit too complex. Could you try to refactor it into a few protected methods instead of performing all the logic here. If you think that’s too much, could you at least move everything within the if-clause into another method. Thanks in advance!

pyreweb2 committed:
Update composer.json

fix indent

pyreweb2 committed:
Splitting wildcard domain check in fonctions

And even refactored a little bit to use preg_grep instead of testing against a string and filtering out after.

hugovk commented on src/CorsService.php:
```diff
@@ -358,46 +358,64 @@ protected function isOriginAllowed($origin)
+	    if ($this->isOriginInWildcardOrigins($origin)) {
+	        return true;
+	    }
+        }
```
Nit: align indents.

Tip: use an editor on your computer rather than the GitHub editor.

pyreweb2 replied:
I'll try to use git correctly instead of copy-pasting, yes...

hugovk commented on src/CorsService.php:
```diff
@@ -452,7 +470,26 @@ protected function isAllOriginsAllowed()
+    {
+        $matchedOrigins = $this->getWildcardOrigins();
+	
+	return (count($matchedOrigins) > 0);
```
Nit: indents

hugovk commented on src/CorsService.php:
```diff
@@ -358,46 +358,64 @@ protected function isOriginAllowed($origin)
+        $matchedOrigins = $this->getWildcardOrigins();
+        $hasMatch = false;
+
+        //Lets parse the uri to extract some parts
```
Nits: space after `//` (here and below), "Let's", "URI"

hugovk commented on composer.json:
```diff
@@ -10,7 +10,8 @@
     "require": {
-        "php": ">=5.6"
+        "php": ">=5.6",
+	"league/uri": "4.2.3"
```
Nit: indents

Commit:
Idents fix

Fix the GitHub editor mess

Pyreweb commented:
I'll hope it's ok this time.
Thanks for your patience.

Pyreweb commented:
Hi,
It seems you forgot to accept my PR, now that's it's close to the project standard ?

cniska commented on src/CorsService.php:
```diff
@@ -414,7 +470,26 @@ protected function isAllOriginsAllowed()
+    {
+        $matchedOrigins = $this->getWildcardOrigins();
+        
+        return (count($matchedOrigins) > 0);
```
No need for parenthesis, it’s one expression.

pyreweb2 replied:
True, it fact its only there to help me read and evaluate them, I get it as an implicit bool cast.

cniska commented on src/CorsService.php:
```diff
@@ -356,10 +357,65 @@ protected function isOriginAllowed($origin)
+            // UriParser consider (rightly) *. as an invalid subdomain, so we have to replace it
+            $wildcardDomain = str_replace('*', 'wildcard', $domain);
+            $domainParsed = $parser->parse($wildcardDomain);
+            if ($domainParsed['scheme'] == $originParsed['scheme']) {
```
Why not use a strict check here?

cniska commented on src/CorsService.php:
```diff
@@ -356,10 +357,65 @@ protected function isOriginAllowed($origin)
+                $domainHostRE = str_replace('*\.', '(.+\.)?', $domainHostRE);
+
+                if (preg_match('#'.$domainHostRE.'#', $originParsed['host']) !== 0) {
+                    $hasMatch = true;
```
I would reverse the logic here for early return instead. What I mean is return true here and false at the end. There is no point in keep looping if we already found a match, right?

pyreweb2 replied:
I usually don't like to have return in the middle of loops.
But yes, why continue to iterate when we have a match ? I'm changing it !

cniska commented on src/CorsService.php:
```diff
@@ -414,7 +470,26 @@ protected function isAllOriginsAllowed()
+    protected function getWildcardOrigins()
+    {
+        // We'll search for an *. in all the allowed origins
+        return preg_grep('#\*\.#', $this->allowOrigins);
```
This should be explained inside the docblock instead.

cniska commented on src/CorsService.php:
```diff
@@ -356,10 +357,65 @@ protected function isOriginAllowed($origin)
+                // Place the RegExp in the string
+                $domainHostRE = str_replace('*\.', '(.+\.)?', $domainHostRE);
+
+                if (preg_match('#'.$domainHostRE.'#', $originParsed['host']) !== 0) {
```
I don’t know if it’s just me, but the logic to get here still seem far more complex than it would have to be. @Jalle19 what do you think?

pyreweb2 replied:
If you need any precision or explanation, I'm around

cniska reviewed (changes requested):
I’m sorry, but this still requires some minor fixes.

Commit:
Strict compare strings when checking domain scheme

Commit:
Better docblock for getWildcardOrigins

Commit:
CS fix for hasWildcardOrigins

"No need for parenthesis, it’s one expression."

Commit:
Direct return in isOriginInWildcardOrigins

Why continue to iterate when we have a match ?

Commit:
Adding security $ to isOriginInWildcardOrigins

Without it, that matched my.sudomain.domain.com, sudomain.domain.com, domain.com but also domain.com.evil :/

Jalle19 commented:
I'm going to close this PR due to inactivity. It is still missing test cases for example so it can't be accepted in its current form. Feel free to re-open this PR if you find the time to finish it.
